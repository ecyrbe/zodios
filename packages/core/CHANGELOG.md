# @zodios/core

## 11.0.0-beta.19

### Major Changes

- 583775e: feat(openapi): add openapi

## 11.0.0-beta.18

### Major Changes

- 58016f5: migrate to typescript v11

## 11.0.0-beta.17

### Major Changes

- 48a20cc: fix(react): do not double fetch if react-query signal is not used
- 8303e57: feat(plugin): add new zodios plugin system

## 11.0.0-beta.16

### Major Changes

- f1b38f4: feat: improve typings
- cb56d43: feat: improve typings

## 11.0.0-beta.15

### Major Changes

- eaf8e45: feat(fetcher): change fetcher model to factory
- faaec6f: chore(deps): update zod
- f161353: chore(deps): update fp-ts
- 5115eaa: chore(deps): update typescript
- 4ac4921: feat(react): simplify exported instance

## 11.0.0-beta.14

### Major Changes

- 422221b: refactor: factor http verbs
- 13247c6: chore: fix type exports for nodenext

## 11.0.0-beta.13

### Major Changes

- 2745e1a: feat(react): migrate react to v11
- 98433a9: fix(react): bump version

## 11.0.0-beta.12

### Major Changes

- 29dcf47: fix(testing): safely export hooks

## 11.0.0-beta.11

### Major Changes

- a753f07: fix(testing): import types from core instead of relaive directory
- 33b5de0: refactor(testing): move base hooks to testing

## 11.0.0-beta.10

### Major Changes

- c360c52: feat(testing): add response testing handlers

## 11.0.0-beta.9

### Major Changes

- c4c00a1: chore(deps): upgrade axios
- c6a4d0e: feat(testing): add testing package

## 11.0.0-beta.8

### Major Changes

- e67fa49: feat(fetch): remove qs dependency
- 731e399: tests(core): fix tests

## 11.0.0-beta.7

### Major Changes

- 7dcbb0c: chore(fetch): fix deps

## 11.0.0-beta.6

### Major Changes

- da4ddab: add fetch package
- 535c76b: add basic mocks helpers

## 11.0.0-beta.5

### Major Changes

- 1b67309: fix axios tests
- c7c03bc: sync axios and core versions

## 11.0.0-beta.4

### Major Changes

- 6c10dc2: rename exports
- 679a267: update axios

### Patch Changes

- 14ce00b: add fetch implementation
- c9cae0a: autocommit on changesets
- 3ac25a1: add fetcher provider and axios impl
- 90108a5: body is now part of the config

## 11.0.0-beta.3

### Patch Changes

- add publish script
- 5575e28: add tests for type providers

## 11.0.0-beta.2

### Patch Changes

- remove publish action from ci

## 11.0.0-beta.1

### Patch Changes

- migrate to monorepo
