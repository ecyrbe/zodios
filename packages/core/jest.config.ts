import type { Config } from "@jest/types";

// Objet synchrone
const config: Config.InitialOptions = {
  verbose: true,
  moduleFileExtensions: ["ts", "js", "json", "node"],
  rootDir: "./",
  testRegex: ".(spec|test).tsx?$",
  transform: {
    "^.+\\.tsx?$": "ts-jest",
  },
  coveragePathIgnorePatterns: ["<rootDir>/node_modules/", "index\\.ts"],
  coverageDirectory: "./coverage",
  coverageThreshold: {
    global: {
      branches: 80,
      functions: 85,
      lines: 85,
      statements: 85,
    },
  },
  testEnvironment: "node",
};
export default config;
