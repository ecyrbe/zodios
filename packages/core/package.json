{
  "name": "@zodios/core",
  "description": "Typescript API client with autocompletion and zod validations",
  "version": "11.0.0-beta.19",
  "main": "lib/index.js",
  "module": "lib/index.mjs",
  "typings": "lib/index.d.ts",
  "exports": {
    ".": {
      "import": "./lib/index.mjs",
      "require": "./lib/index.js",
      "types": "./lib/index.d.ts"
    },
    "./lib/*.types": {
      "import": "./lib/*.types.mjs",
      "require": "./lib/*.types.js",
      "types": "./lib/*.types.d.ts"
    },
    "./package.json": "./package.json"
  },
  "files": [
    "lib"
  ],
  "author": {
    "name": "ecyrbe",
    "email": "ecyrbe@gmail.com"
  },
  "homepage": "https://github.com/ecyrbe/zodios",
  "repository": {
    "type": "git",
    "url": "https://github.com/ecyrbe/zodios.git"
  },
  "license": "MIT",
  "keywords": [
    "openapi",
    "zod",
    "autocomplete",
    "validation"
  ],
  "scripts": {
    "prebuild": "rimraf lib",
    "example": "ts-node examples/jsonplaceholder.ts",
    "example:dev.to": "ts-node examples/dev.to/example.ts",
    "build": "tsup",
    "test": "jest --coverage"
  },
  "peerDependencies": {
    "fp-ts": "2.x",
    "io-ts": "2.x",
    "zod": "^3.x"
  },
  "peerDependenciesMeta": {
    "zod": {
      "optional": true
    },
    "fp-ts": {
      "optional": true
    },
    "io-ts": {
      "optional": true
    }
  },
  "devDependencies": {
    "@jest/globals": "29.5.0",
    "@jest/types": "29.5.0",
    "@types/jest": "29.5.0",
    "@types/node": "18.15.11",
    "form-data": "^4.0.0",
    "fp-ts": "2.13.1",
    "io-ts": "2.2.20",
    "jest": "29.5.0",
    "rimraf": "4.4.1",
    "ts-jest": "29.1.0",
    "ts-node": "10.9.1",
    "tsup": "6.7.0",
    "typescript": "5.0.4",
    "zod": "3.21.4"
  }
}
