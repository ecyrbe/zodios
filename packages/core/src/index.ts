export { ZodiosCore } from "./zodios";
export type {
  ApiOf,
  TypeProviderOf,
  FetcherProviderOf,
  ZodiosInstance,
  ZodiosBase,
} from "./zodios";
export { ZodiosError } from "./zodios-error";
export type {
  AnyZodiosMethodOptions,
  AnyZodiosRequestOptions,
  ZodiosBodyForEndpoint,
  ZodiosBodyByPath,
  ZodiosBodyByAlias,
  ZodiosHeaderParamsForEndpoint,
  ZodiosHeaderParamsByPath,
  ZodiosHeaderParamsByAlias,
  Method,
  ZodiosPathParams,
  ZodiosPathParamsForEndpoint,
  ZodiosPathParamsByPath,
  ZodiosPathParamByAlias,
  ZodiosPathsByMethod,
  ZodiosResponseForEndpoint,
  ZodiosResponseByPath,
  ZodiosResponseByAlias,
  ZodiosQueryParamsForEndpoint,
  ZodiosQueryParamsByPath,
  ZodiosQueryParamsByAlias,
  FindZodiosEndpointDefinitionByPath,
  FindZodiosEndpointDefinitionByAlias,
  ZodiosErrorForEndpoint,
  ZodiosErrorByPath,
  ZodiosErrorByAlias,
  ZodiosErrorsByPath,
  ZodiosErrorsByAlias,
  ZodiosEndpointDefinition,
  ZodiosEndpointParameter,
  ZodiosEndpointError,
  ZodiosOptions,
  ZodiosRequestOptions,
  ZodiosRequestOptionsByPath,
  ZodiosRequestOptionsByAlias,
  ZodiosPlugin,
} from "./zodios.types";
export {
  HTTP_METHODS,
  HTTP_MUTATION_METHODS,
  HTTP_QUERY_METHODS,
} from "./zodios.types";
export type {
  AnyZodiosTypeProvider,
  InferInputTypeFromSchema,
  InferOutputTypeFromSchema,
  IoTsTypeProvider,
  TsTypeProvider,
  ZodiosRuntimeTypeProvider,
  ZodiosValidateResult,
  ZodTypeProvider,
} from "./type-providers";
export {
  ioTsTypeProvider,
  tsTypeProvider,
  zodTypeProvider,
  tsSchema,
  tsFnSchema,
} from "./type-providers";
export type {
  AnyZodiosFetcherProvider,
  TypeOfFetcherConfig,
  TypeOfFetcherError,
  TypeOfFetcherOptions,
  TypeOfFetcherResponse,
  ZodiosFetcherFactoryOptions,
  ZodiosFetcherFactory,
  ZodiosFetcher,
} from "./fetcher-providers";
export { setFetcherHook, clearFetcherHook } from "./hooks";
export type { PluginId } from "./plugins";
export {
  schemaValidationPlugin,
  formDataPlugin,
  formURLPlugin,
  headerPlugin,
} from "./plugins";

export {
  makeApi,
  apiBuilder,
  parametersBuilder,
  makeParameters,
  makeEndpoint,
  makeErrors,
  checkApi,
  prefixApi,
  mergeApis,
} from "./api";
