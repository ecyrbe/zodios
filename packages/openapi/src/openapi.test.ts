import { makeApi } from "@zodios/core";
import { z } from "zod";
import {
  basicAuthScheme,
  bearerAuthScheme,
  oauth2Scheme,
  openApiBuilder,
} from "./openapi";

const user = z.object({
  id: z.string(),
  name: z.string(),
  email: z.string().email(),
});

const api = makeApi([
  {
    method: "get",
    path: "/users?filter=:filter#fragment",
    alias: "getUsers",
    description: "Get all users",
    parameters: [
      {
        name: "limit",
        type: "Query",
        description: "Limit the number of users",
        schema: z.number().positive().default(10),
      },
      {
        name: "offset",
        type: "Query",
        description: "Offset the number of users",
        schema: z.number().positive().optional(),
      },
      {
        name: "filter",
        type: "Query",
        description: "Filter users by name",
        schema: z
          .array(z.string())
          .refine((a) => new Set(a).size === a.length, "No duplicates allowed"),
      },
    ],
    response: z.array(user),
    errors: [
      {
        status: 404,
        description: "No users found",
        schema: z.object({
          message: z.literal("No users found"),
        }),
      },
      {
        status: "default",
        description: "Default error",
        schema: z.object({
          message: z.string(),
        }),
      },
    ],
  },
  {
    method: "get",
    path: "/users/:id",
    alias: "getUser",
    description: "Get a user by id",
    response: user,
  },
  {
    method: "post",
    path: "/users",
    alias: "createUser",
    description: "Create a user",
    parameters: [
      {
        name: "user",
        type: "Body",
        description: "The user to create",
        schema: user.omit({ id: true }),
      },
    ],
    response: user,
  },
  {
    method: "put",
    path: "/users/:id",
    alias: "updateUser",
    description: "Update a user",
    parameters: [
      {
        name: "user",
        type: "Body",
        description: "The user to update",
        schema: user,
      },
    ],
    response: user,
  },
  {
    method: "delete",
    path: "/users/:id",
    alias: "deleteUser",
    description: "Delete a user",
    response: z.void(),
    status: 204,
  },
]);

describe("toOpenApi", () => {
  it("should generate bearer scheme", () => {
    const scheme = bearerAuthScheme();
    expect(scheme).toEqual({
      type: "http",
      scheme: "bearer",
      bearerFormat: "JWT",
    });
  });

  it("should generate basic scheme", () => {
    const scheme = basicAuthScheme();
    expect(scheme).toEqual({
      type: "http",
      scheme: "basic",
    });
  });

  it("should generate oauth2 scheme", () => {
    const scheme = oauth2Scheme({
      implicit: {
        authorizationUrl: "https://example.com/oauth2/authorize",
        scopes: {
          read: "Read access",
          write: "Write access",
        },
      },
    });
    expect(scheme).toEqual({
      type: "oauth2",
      flows: {
        implicit: {
          authorizationUrl: "https://example.com/oauth2/authorize",
          scopes: {
            read: "Read access",
            write: "Write access",
          },
        },
      },
    });
  });

  it("should convert to openapi with builder", () => {
    const openApi = openApiBuilder({
      title: "My API",
      version: "1.0.0",
    })
      .addPublicApi(api)
      .build();
    expect(openApi).toMatchSnapshot();
  });

  it("should convert to openapi with builder with security", () => {
    const openApi = openApiBuilder({
      title: "My API",
      version: "1.0.0",
    })
      .addServer({ url: "/api/v1" })
      .addSecurityScheme("auth", bearerAuthScheme())
      .addProtectedApi("auth", api)
      .build();
    expect(openApi).toMatchSnapshot();
  });
});
