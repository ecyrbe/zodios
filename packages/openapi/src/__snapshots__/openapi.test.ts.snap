// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toOpenApi should convert to openapi with builder 1`] = `
{
  "info": {
    "title": "My API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/users": {
      "post": {
        "description": "Create a user",
        "operationId": "createUser",
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "email": {
                    "format": "email",
                    "type": "string",
                  },
                  "name": {
                    "type": "string",
                  },
                },
                "required": [
                  "name",
                  "email",
                ],
                "type": "object",
              },
            },
          },
          "description": "The user to create",
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "email": {
                      "format": "email",
                      "type": "string",
                    },
                    "id": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                    "email",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
        "security": undefined,
        "summary": "Create a user",
        "tags": [
          "users",
        ],
      },
    },
    "/users/{id}": {
      "delete": {
        "description": "Delete a user",
        "operationId": "deleteUser",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "requestBody": undefined,
        "responses": {
          "204": {
            "content": {
              "application/json": {
                "schema": {},
              },
            },
            "description": "Success",
          },
        },
        "security": undefined,
        "summary": "Delete a user",
        "tags": [
          "users",
        ],
      },
      "get": {
        "description": "Get a user by id",
        "operationId": "getUser",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "requestBody": undefined,
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "email": {
                      "format": "email",
                      "type": "string",
                    },
                    "id": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                    "email",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
        "security": undefined,
        "summary": "Get a user by id",
        "tags": [
          "users",
        ],
      },
      "put": {
        "description": "Update a user",
        "operationId": "updateUser",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "email": {
                    "format": "email",
                    "type": "string",
                  },
                  "id": {
                    "type": "string",
                  },
                  "name": {
                    "type": "string",
                  },
                },
                "required": [
                  "id",
                  "name",
                  "email",
                ],
                "type": "object",
              },
            },
          },
          "description": "The user to update",
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "email": {
                      "format": "email",
                      "type": "string",
                    },
                    "id": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                    "email",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
        "security": undefined,
        "summary": "Update a user",
        "tags": [
          "users",
        ],
      },
    },
    "/users?filter={filter}#fragment": {
      "get": {
        "description": "Get all users",
        "operationId": "getUsers",
        "parameters": [
          {
            "in": "path",
            "name": "filter",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
          {
            "description": "Limit the number of users",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "default": 10,
              "exclusiveMinimum": true,
              "minimum": 0,
              "type": "number",
            },
          },
          {
            "description": "Offset the number of users",
            "in": "query",
            "name": "offset",
            "required": false,
            "schema": {
              "exclusiveMinimum": true,
              "minimum": 0,
              "type": "number",
            },
          },
          {
            "description": "Filter users by name",
            "in": "query",
            "name": "filter[]",
            "required": true,
            "schema": {
              "items": {
                "type": "string",
              },
              "type": "array",
            },
          },
        ],
        "requestBody": undefined,
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "items": {
                    "additionalProperties": false,
                    "properties": {
                      "email": {
                        "format": "email",
                        "type": "string",
                      },
                      "id": {
                        "type": "string",
                      },
                      "name": {
                        "type": "string",
                      },
                    },
                    "required": [
                      "id",
                      "name",
                      "email",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
              },
            },
            "description": "Success",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "message": {
                      "enum": [
                        "No users found",
                      ],
                      "type": "string",
                    },
                  },
                  "required": [
                    "message",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "No users found",
          },
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "message",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default error",
          },
        },
        "security": undefined,
        "summary": "Get all users",
        "tags": [
          "users",
        ],
      },
    },
  },
  "servers": undefined,
}
`;

exports[`toOpenApi should convert to openapi with builder with security 1`] = `
{
  "components": {
    "securitySchemes": {
      "auth": {
        "bearerFormat": "JWT",
        "description": undefined,
        "scheme": "bearer",
        "type": "http",
      },
    },
  },
  "info": {
    "title": "My API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/users": {
      "post": {
        "description": "Create a user",
        "operationId": "createUser",
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "email": {
                    "format": "email",
                    "type": "string",
                  },
                  "name": {
                    "type": "string",
                  },
                },
                "required": [
                  "name",
                  "email",
                ],
                "type": "object",
              },
            },
          },
          "description": "The user to create",
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "email": {
                      "format": "email",
                      "type": "string",
                    },
                    "id": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                    "email",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
        "security": [
          {
            "auth": [],
          },
        ],
        "summary": "Create a user",
        "tags": [
          "users",
        ],
      },
    },
    "/users/{id}": {
      "delete": {
        "description": "Delete a user",
        "operationId": "deleteUser",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "requestBody": undefined,
        "responses": {
          "204": {
            "content": {
              "application/json": {
                "schema": {},
              },
            },
            "description": "Success",
          },
        },
        "security": [
          {
            "auth": [],
          },
        ],
        "summary": "Delete a user",
        "tags": [
          "users",
        ],
      },
      "get": {
        "description": "Get a user by id",
        "operationId": "getUser",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "requestBody": undefined,
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "email": {
                      "format": "email",
                      "type": "string",
                    },
                    "id": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                    "email",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
        "security": [
          {
            "auth": [],
          },
        ],
        "summary": "Get a user by id",
        "tags": [
          "users",
        ],
      },
      "put": {
        "description": "Update a user",
        "operationId": "updateUser",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "email": {
                    "format": "email",
                    "type": "string",
                  },
                  "id": {
                    "type": "string",
                  },
                  "name": {
                    "type": "string",
                  },
                },
                "required": [
                  "id",
                  "name",
                  "email",
                ],
                "type": "object",
              },
            },
          },
          "description": "The user to update",
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "email": {
                      "format": "email",
                      "type": "string",
                    },
                    "id": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                    "email",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
        "security": [
          {
            "auth": [],
          },
        ],
        "summary": "Update a user",
        "tags": [
          "users",
        ],
      },
    },
    "/users?filter={filter}#fragment": {
      "get": {
        "description": "Get all users",
        "operationId": "getUsers",
        "parameters": [
          {
            "in": "path",
            "name": "filter",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
          {
            "description": "Limit the number of users",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "default": 10,
              "exclusiveMinimum": true,
              "minimum": 0,
              "type": "number",
            },
          },
          {
            "description": "Offset the number of users",
            "in": "query",
            "name": "offset",
            "required": false,
            "schema": {
              "exclusiveMinimum": true,
              "minimum": 0,
              "type": "number",
            },
          },
          {
            "description": "Filter users by name",
            "in": "query",
            "name": "filter[]",
            "required": true,
            "schema": {
              "items": {
                "type": "string",
              },
              "type": "array",
            },
          },
        ],
        "requestBody": undefined,
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "items": {
                    "additionalProperties": false,
                    "properties": {
                      "email": {
                        "format": "email",
                        "type": "string",
                      },
                      "id": {
                        "type": "string",
                      },
                      "name": {
                        "type": "string",
                      },
                    },
                    "required": [
                      "id",
                      "name",
                      "email",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
              },
            },
            "description": "Success",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "message": {
                      "enum": [
                        "No users found",
                      ],
                      "type": "string",
                    },
                  },
                  "required": [
                    "message",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "No users found",
          },
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "message",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default error",
          },
        },
        "security": [
          {
            "auth": [],
          },
        ],
        "summary": "Get all users",
        "tags": [
          "users",
        ],
      },
    },
  },
  "servers": [
    {
      "url": "/api/v1",
    },
  ],
}
`;
